File: apple/AestheticAlgorithm/Algorithms/Sorting/SelectionSort.swift
import Foundation

struct SelectionSort: AlgorithmExecutable {
    let id = "selection-sort"
    let name = "Selection Sort"
    let category = "sorting"
    let timeComplexity = (best: "O(n²)", average: "O(n²)", worst: "O(n²)")
    let spaceComplexity = "O(1)"

    let pseudocode: [(line: Int, text: String)] = [
        (0, "procedure selectionSort(A: list)"),
        (1, "  n ← length(A)"),
        (2, "  for i ← 0 to n - 2 do"),
        (3, "    minIndex ← i"),
        (4, "    for j ← i + 1 to n - 1 do"),
        (5, "      if A[j] < A[minIndex] then"),
        (6, "        minIndex ← j"),
        (7, "    if minIndex ≠ i then"),
        (8, "      swap(A[i], A[minIndex])"),
        (9, "  return A")
    ]

    func generateSteps(from input: [Int]) -> [AlgorithmStep] {
        var array = input
        var steps: [AlgorithmStep] = []
        let n = array.count
        var sortedIndices: [Int] = []

        for i in 0..<(n - 1) {
            var minIndex = i

            // Highlight the current position and initial min candidate
            steps.append(AlgorithmStep(
                type: .select,
                array: array,
                highlightedIndices: [minIndex],
                secondaryIndices: [],
                sortedIndices: sortedIndices,
                pseudocodeLine: 3,
                description: "Setting minIndex ← \(i), current minimum is \(array[i])"
            ))

            // Scan through unsorted portion for minimum
            for j in (i + 1)..<n {
                // Compare current element with current minimum
                steps.append(AlgorithmStep(
                    type: .compare,
                    array: array,
                    highlightedIndices: [j, minIndex],
                    secondaryIndices: [],
                    sortedIndices: sortedIndices,
                    pseudocodeLine: 5,
                    description: "Comparing \(array[j]) with current minimum \(array[minIndex])"
                ))

                if array[j] < array[minIndex] {
                    minIndex = j
                    // New minimum found
                    steps.append(AlgorithmStep(
                        type: .select,
                        array: array,
                        highlightedIndices: [minIndex],
                        secondaryIndices: [i],
                        sortedIndices: sortedIndices,
                        pseudocodeLine: 6,
                        description: "New minimum found: \(array[minIndex]) at index \(minIndex)"
                    ))
                }
            }

            // Swap minimum to its final position if needed
            if minIndex != i {
                steps.append(AlgorithmStep(
                    type: .swap,
                    array: array,
                    highlightedIndices: [i, minIndex],
                    secondaryIndices: [],
                    sortedIndices: sortedIndices,
                    pseudocodeLine: 8,
                    description: "Swapping \(array[i]) at index \(i) with \(array[minIndex]) at index \(minIndex)"
                ))
                array.swapAt(i, minIndex)
            }

            // Mark position i as sorted
            sortedIndices.append(i)
            steps.append(AlgorithmStep(
                type: .sorted,
                array: array,
                highlightedIndices: [i],
                secondaryIndices: [],
                sortedIndices: sortedIndices,
                pseudocodeLine: 2,
                description: "\(array[i]) is now in its final sorted position at index \(i)"
            ))
        }

        // The last element is sorted by default
        sortedIndices.append(n - 1)
        steps.append(AlgorithmStep(
            type: .sorted,
            array: array,
            highlightedIndices: [n - 1],
            secondaryIndices: [],
            sortedIndices: sortedIndices,
            pseudocodeLine: 9,
            description: "Last element \(array[n - 1]) is in its final position. Array is sorted."
        ))

        return steps
    }
}

File: web/src/algorithms/sorting/selection-sort.ts
import type { AlgorithmImplementation, AlgorithmStep } from '@/lib/types/algorithm';

export const selectionSort: AlgorithmImplementation = {
  id: 'selection-sort',
  name: 'Selection Sort',
  category: 'sorting',
  timeComplexity: { best: 'O(n²)', average: 'O(n²)', worst: 'O(n²)' },
  spaceComplexity: 'O(1)',
  pseudocode: [
    { line: 0, text: 'procedure selectionSort(A: list)' },
    { line: 1, text: '  n ← length(A)' },
    { line: 2, text: '  for i ← 0 to n - 2 do' },
    { line: 3, text: '    minIdx ← i' },
    { line: 4, text: '    for j ← i + 1 to n - 1 do' },
    { line: 5, text: '      if A[j] < A[minIdx] then' },
    { line: 6, text: '        minIdx ← j' },
    { line: 7, text: '    swap(A[i], A[minIdx])' },
    { line: 8, text: '  return A' },
  ],

  generateSteps(input: number[]): AlgorithmStep[] {
    const array = [...input];
    const steps: AlgorithmStep[] = [];
    const n = array.length;
    const sortedIndices: number[] = [];

    for (let i = 0; i < n - 1; i++) {
      let minIdx = i;

      // Select current position
      steps.push({
        type: 'select',
        array: [...array],
        highlightedIndices: [i],
        secondaryIndices: [],
        sortedIndices: [...sortedIndices],
        pseudocodeLine: 3,
        description: `Finding minimum in unsorted portion starting at index ${i}`,
      });

      for (let j = i + 1; j < n; j++) {
        // Compare current element with current minimum
        steps.push({
          type: 'compare',
          array: [...array],
          highlightedIndices: [j, minIdx],
          secondaryIndices: [i],
          sortedIndices: [...sortedIndices],
          pseudocodeLine: 5,
          description: `Comparing ${array[j]} with current min ${array[minIdx]}`,
        });

        if (array[j] < array[minIdx]) {
          minIdx = j;

          // New minimum found
          steps.push({
            type: 'select',
            array: [...array],
            highlightedIndices: [minIdx],
            secondaryIndices: [i],
            sortedIndices: [...sortedIndices],
            pseudocodeLine: 6,
            description: `New minimum found: ${array[minIdx]} at index ${minIdx}`,
          });
        }
      }

      // Swap minimum to its final position
      if (minIdx !== i) {
        [array[i], array[minIdx]] = [array[minIdx], array[i]];

        steps.push({
          type: 'swap',
          array: [...array],
          highlightedIndices: [i, minIdx],
          secondaryIndices: [],
          sortedIndices: [...sortedIndices],
          pseudocodeLine: 7,
          description: `Swapping ${array[i]} into position ${i}`,
        });
      }

      // Mark position as sorted
      sortedIndices.push(i);
      steps.push({
        type: 'sorted',
        array: [...array],
        highlightedIndices: [i],
        secondaryIndices: [],
        sortedIndices: [...sortedIndices],
        pseudocodeLine: 7,
        description: `${array[i]} is in its final position`,
      });
    }

    // Last element is automatically sorted
    sortedIndices.push(n - 1);
    steps.push({
      type: 'sorted',
      array: [...array],
      highlightedIndices: [n - 1],
      secondaryIndices: [],
      sortedIndices: [...sortedIndices],
      pseudocodeLine: 8,
      description: `${array[n - 1]} is in its final position`,
    });

    return steps;
  },
};
